Add ChangeLog

yan12125 opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Fixes #1626.

TRox1972 commented:
Any good way to do this automatically, e.g. with git?

yan12125 commented:
Do you have some commands for that? I hope to have something more readable than ```git log --oneline``` or ```git shortlog```.

TRox1972 commented:
Found [this](https://github.com/vaab/gitchangelog) after a quick search

yan12125 commented:
@phihag @dstftw @jaimeMF I guess it's OK to have external dependencies in developer-only scripts?

TRox1972 commented:
[This](https://github.com/skywinder/github-changelog-generator) is another alternative. It's more specified towards GitHub (using their API), though it it is written in Ruby.

dstftw commented:
Why CamelCase?
If you plan this to be auto generated from git log then commit messages should be very accurate, clean and isolated that is barely possible in our case.
I'm fine with external dependencies if they work across major platforms.

yan12125 commented:
> Why CamelCase?

I follow the convention of GNU autotools [1]. Other names like [Changelog](https://github.com/FFmpeg/FFmpeg/blob/master/Changelog) or [NEWS](https://hg.python.org/cpython/file/tip/Misc/NEWS) are also fine.

> commit messages should be very accurate, clean and isolated

You're right. In fact I think commit message are for developers. For example I guess ```[compat] Fix compat_shlex_split for non-ASCII input``` is clear and accurate for developers, while users may not be able to get it in a short time.

I'm not sure whether maintaining an additional change log is a burden or not, so automation is always in my mind.

[1] https://www.gnu.org/software/automake/manual/automake.html

yan12125 commented:
Agreed with @dstftw - generating a changelog from commit messages is infeasible for youtube-dl. Just pushed the latest changelog. Any comments on the format or other things?

dstftw commented:
1. When do you expect changes in changelog to be applied: after each relevant commit or during the release process? If latter then releaser may easily overlook some changes. In any case release scripts should be tweaked to at least insert a version number.
2. What's origin of suggested changelog layout? Such huge indents look very odd.

yan12125 commented:
> When do you expect changes in changelog to be applied

For commits by collaborators, include modifications to the changelog in each relevant commit. An issue is pull request authors. If they include changes to the changelog, there will be lots of merge conflicts. An idea is the one that merges a pull request should add some notes to the changelog.

> In any case release scripts should be tweaked to at least insert a version number.

Added to ```devscripts/release.sh```

> What's origin of suggested changelog layout? Such huge indents look very odd.

Well, vim's ```changelog.vim``` forces real tabs (```\t```). Sorry for not noticing that. Updated.

dstftw commented:
Well, you did not get it. What's the point of indents at all? As for me it's just waste of space and unnecessary margins.

TRox1972 commented:
What about markdown instead of indents?

dstftw commented:
Markdown does not make much sense for files that should be distributes plain text.

TRox1972 commented:
OK. Should changes be categorized within each release, or should there just be a list of the changes, e.g. like [FFmpeg's](https://github.com/FFmpeg/FFmpeg/blob/master/Changelog)?

yan12125 commented:
Updated with eliminating leading spaces and trailing colons in group titles.

@TRox1972 I like grouped ChangeLog more. For example Python's [NEWS](https://hg.python.org/cpython/raw-file/default/Misc/NEWS) is more readable than FFMpeg's [Changelog](https://github.com/FFmpeg/FFmpeg/blob/master/Changelog) for me.

dstftw commented on ChangeLog:
```diff
@@ -0,0 +1,141 @@
+- rai (#8617, #9157, #9232, #8552, #8551)
+- nationalgeographic (#9991)
+- iqiyi
+- Miscellaneous
```
I guess such ambiguous notes should not be included in changelog.

yan12125 replied:
Sorry it's a typo ```- Miscellaneous``` -> ```Miscellaneous```

dstftw commented on ChangeLog:
```diff
@@ -0,0 +1,141 @@
+- iqiyi
+- Miscellaneous
+- Add coding convention guide to README.md
+- Add PULL_REQUEST_TEMPLATE.md
```
Is there any point including changes directly not related to youtube-dl executable?

yan12125 replied:
For me ChangeLog is for the whole project, so it may contain anything related to this project. If you think such changes are not important enough to be in ChangeLog, I'll remove them.

dstftw commented:
It's probably reasonable to use different prefixes for different changes, e.g. `+` for additions, `-` for deletions and `*` for regular changes, etc. Some examples:
```
+ hrti extractor (#9482)
+ Added bbc extractor
+ New extractors: // unsure about inner lists
 + hrti (#9482)
 + carambatv
+ New cool feature `--blah`
- bbc extractor

* Fixed/improved extractors:
 * theplatform
 * aenetworks
 * nationalgeographic
 * hrti (#9482)
 * facebook (#5701)
 * buzzfeed (#5701)
 * rai (#8617, #9157, #9232, #8552, #8551)
 * nationalgeographic (#9991)
 * iqiyi

* Fixed format selection with filters.
```

yan12125 commented:
Further formatting is fine for humans but not machines. I guess people may use the raw ChangeLog to generate different formats (HTML, Markdown, etc.), so I hope ChangeLog can take a balance between human-readable and machine-readable.

dstftw commented:
Changelog should also be bundled in `youtube-dl.tar.gz`.

yan12125 commented:
> Changelog should also be bundled in ```youtube-dl.tar.gz```.

Thanks. Done.

dstftw commented on ChangeLog:
```diff
@@ -0,0 +1,237 @@
+version <unreleased>
```
Isn't it easier to just prepend the actual string during release rather than remember to always keep `version <unreleased>` on top? Also for substitution pattern we usually use `@UPPERCASE@`.

yan12125 replied:
Some people may be interested in unreleased changes as well. I guess ```version <unreleased>``` is more human-readable than ```version %UNRELEASED%```.

My idea comes primarily from [FFmpeg](https://github.com/FFmpeg/FFmpeg/blob/master/Changelog).

dstftw commented on ChangeLog:
```diff
@@ -0,0 +1,237 @@
+- cbc
+- mgtv
+- orf
+
```
With two blank lines between versions it's easier to visually distinguish where one ends and another starts.

yan12125 replied:
Like this?
```
+- shared (#10170)
+- soundcloud (#10179)
+- twitch (#9767)
+
+
+version 2016.07.26.2
+
+Fixed/improved extractors
+- smotri
+- camdemy
+- mtv
+- comedycentral
+- cmt
+- cbc
+- mgtv
+- orf
+
```

dstftw replied:
Yes.

yan12125 replied:
Done.

dstftw commented:
LGTM.

yan12125 commented:
Rebased and pushed to 35aa6c538f593eb47169e6b06ef383ad12b45930.
